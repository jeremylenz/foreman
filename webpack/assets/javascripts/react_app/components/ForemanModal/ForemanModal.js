import React from 'react';
import { Modal } from 'patternfly-react';
import PropTypes from 'prop-types';
import ForemanModalHeader from './subcomponents/ForemanModalHeader';
import ForemanModalFooter from './subcomponents/ForemanModalFooter';
import ModalContext from './ForemanModalContext';
import { extractModalNodes } from './helpers';

const ForemanModal = props => {
  const { isOpen, onClose, title, ...propsToPassDown } = props; // don't pass down those two as it causes PF problems
  // Extract header and footer from children, if provided
  const { headerChild, footerChild, otherChildren } = extractModalNodes(
    props.children
  );
  const context = {
    isOpen,
    onClose,
    title,
  };

  return (
    <ModalContext.Provider value={context}>
      {/* Change the name of the props we are passing down to Modal to conform to Patternfly 3 api */}
      <Modal
        onHide={onClose}
        show={isOpen}
        className="foreman-modal"
        {...propsToPassDown}
      >
        {headerChild}
        <Modal.Body>{otherChildren}</Modal.Body>
        {footerChild}
      </Modal>
    </ModalContext.Provider>
  );
};

// Header and Footer use the provided children, or default markup if none provided

ForemanModal.Header = ForemanModalHeader;
ForemanModal.Footer = ForemanModalFooter;

ForemanModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  children: PropTypes.node,
  title: PropTypes.string.isRequired,
  onClose: PropTypes.func.isRequired,
};

ForemanModal.defaultProps = {
  children: null,
};

export default ForemanModal;
