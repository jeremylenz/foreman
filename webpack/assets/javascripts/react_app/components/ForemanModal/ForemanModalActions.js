import {
  ADD_MODAL,
  SET_MODAL_OPEN,
  SET_MODAL_CLOSED,
} from './ForemanModalConstants';

// In function components, no need to use these-- just use useDispatch and dispatch the actions directly.
// For class components, these are provided so you can plug them into bindActionCreators

export const addModal = ({ id, open = false }) => (dispatch, getState) => {
  const { foremanModals } = getState();
  if (foremanModals && foremanModals[id]) {
    throw new Error(`ForemanModal with ID ${id} already exists`);
  }
  return dispatch({
    type: ADD_MODAL,
    payload: { id, open },
  });
};

export const setModalOpen = ({ id }) => (dispatch, getState) => {
  const { foremanModals } = getState();
  if (foremanModals && !foremanModals[id]) {
    throw new Error(
      `SET_MODAL_OPEN error: Modal with id '${id}' does not exist`
    );
  }
  return dispatch({
    type: SET_MODAL_OPEN,
    payload: { id },
  });
};

export const setModalClosed = ({ id }) => (dispatch, getState) => {
  const { foremanModals } = getState();
  if (foremanModals && !foremanModals[id]) {
    throw new Error(
      `SET_MODAL_CLOSED error: Modal with id '${id}' does not exist`
    );
  }
  return dispatch({
    type: SET_MODAL_CLOSED,
    payload: { id },
  });
};

// Pass in the ForemanModal id here and get bound action creators with the id already plugged in.
export const bindForemanModalActionsToId = ({ id }) => ({
  addModal: () => addModal({ id }),
  setModalOpen: () => setModalOpen({ id }),
  setModalClosed: () => setModalClosed({ id }),
});
