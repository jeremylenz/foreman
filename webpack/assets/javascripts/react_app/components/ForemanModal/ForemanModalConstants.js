export const SET_MODAL_OPEN = 'SET_MODAL_OPEN';
export const SET_MODAL_CLOSED = 'SET_MODAL_CLOSED';
export const ADD_MODAL = 'ADD_MODAL';
