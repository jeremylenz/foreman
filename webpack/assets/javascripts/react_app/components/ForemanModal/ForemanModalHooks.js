import { useContext } from 'react';
import ModalContext from './ForemanModalContext';

// Because enzyme doesn't support useContext yet
export const useModalContext = () => useContext(ModalContext);

// to get enzyme hacky test to work
export default ModalContext;
