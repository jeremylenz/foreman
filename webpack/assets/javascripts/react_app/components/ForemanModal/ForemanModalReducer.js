import Immutable from 'seamless-immutable';
import {
  SET_MODAL_OPEN,
  SET_MODAL_CLOSED,
  ADD_MODAL,
} from './ForemanModalConstants';

const initialState = Immutable({});

// Modals state has id as key and open state as value:
// { myModal: {open: true} }
// Since keys cannot be duplicated, we avoid creating duplicate modals in this way.

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_MODAL_OPEN: {
      const newState = {};
      newState[action.payload.id] = {
        open: true,
      };
      return state.merge(newState);
    }
    case SET_MODAL_CLOSED: {
      const newState = {};
      newState[action.payload.id] = {
        open: false,
      };
      return state.merge(newState);
    }
    case ADD_MODAL: {
      if (state[action.payload.id]) return state; // if it already exists, don't change its state
      const newState = {};
      newState[action.payload.id] = {
        open: action.payload.open || false,
      };
      return state.merge(newState);
    }
    default:
      return state;
  }
};
