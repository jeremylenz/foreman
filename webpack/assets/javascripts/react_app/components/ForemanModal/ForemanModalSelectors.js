export const selectForemanModalsState = state => state.foremanModals;
export const selectModalStateById = (state, id) =>
  state.foremanModals && state.foremanModals[id];
