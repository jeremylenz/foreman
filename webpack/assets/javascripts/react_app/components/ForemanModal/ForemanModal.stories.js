import React, { useState } from 'react';
import { Button } from 'patternfly-react';
import { storiesOf } from '@storybook/react';
import ForemanModal from '.';
import Story from '../../../../../stories/components/Story';

// This custom Hook is only for the Storybook
const useModalState = initialModalState => {
  const [modalOpen, setModalOpen] = useState(initialModalState);
  const toggleModal = () => setModalOpen(!modalOpen);
  return [modalOpen, toggleModal];
};

storiesOf('Components/ForemanModal', module).add(
  'With default header & footer',
  () =>
    // using createElement here so that hooks work in stories
    React.createElement(() => {
      const [modalOpen, toggleModal] = useModalState(false);
      return (
        <Story>
          <Button bsStyle="primary" onClick={toggleModal}>
            Show Modal
          </Button>
          <ForemanModal
            isOpen={modalOpen}
            title="I'm a modal!"
            onClose={toggleModal}
          >
            <ForemanModal.Header />
            This is the modal body
            <ForemanModal.Footer />
          </ForemanModal>
        </Story>
      );
    })
);

storiesOf('Components/ForemanModal', module).add(
  'With custom header & footer',
  () =>
    React.createElement(() => {
      const [modalOpen, toggleModal] = useModalState(false);
      return (
        <Story>
          <Button bsStyle="primary" onClick={toggleModal}>
            Show Modal
          </Button>
          <ForemanModal
            isOpen={modalOpen}
            title="I'm a custom modal!"
            onClose={toggleModal}
          >
            <ForemanModal.Header>
              <h3>This is a custom header! :)</h3>
            </ForemanModal.Header>
            If a {`<ForemanModal.Header>`} is provided AND has children, the
            title prop will be ignored.
            <ForemanModal.Footer>
              Click the X in the upper right to close
            </ForemanModal.Footer>
          </ForemanModal>
        </Story>
      );
    })
);

storiesOf('Components/ForemanModal', module).add('With no close button', () =>
  React.createElement(() => {
    const [modalOpen, toggleModal] = useModalState(false);
    return (
      <Story>
        <Button bsStyle="primary" onClick={toggleModal}>
          Show Modal
        </Button>
        <ForemanModal
          isOpen={modalOpen}
          title="I'm an X-less modal!"
          onClose={toggleModal}
        >
          <ForemanModal.Header closeButton={false} />
          <br />
          Props passed to ForemanModal.Header will be passed down to
          Modal.Header
          <ForemanModal.Footer />
        </ForemanModal>
      </Story>
    );
  })
);

storiesOf('Components/ForemanModal', module).add('With no footer', () =>
  React.createElement(() => {
    const [modalOpen, toggleModal] = useModalState(false);
    return (
      <Story>
        <Button bsStyle="primary" onClick={toggleModal}>
          Show Modal
        </Button>
        <ForemanModal
          isOpen={modalOpen}
          title="I'm a modal!"
          onClose={toggleModal}
        >
          <ForemanModal.Header />
          This is the modal body. There is no footer.
        </ForemanModal>
      </Story>
    );
  })
);

storiesOf('Components/ForemanModal', module).add('With no header', () =>
  React.createElement(() => {
    const [modalOpen, toggleModal] = useModalState(false);
    return (
      <Story>
        <Button bsStyle="primary" onClick={toggleModal}>
          Show Modal
        </Button>
        <ForemanModal
          isOpen={modalOpen}
          title="I'm a modal!"
          onClose={toggleModal}
        >
          This is the modal body. There is no header.
          <ForemanModal.Footer />
        </ForemanModal>
      </Story>
    );
  })
);

storiesOf('Components/ForemanModal', module).add(
  'With props passed down via spread syntax',
  () =>
    React.createElement(() => {
      const [modalOpen, toggleModal] = useModalState(false);
      return (
        <Story>
          <Button bsStyle="primary" onClick={toggleModal}>
            Show Modal
          </Button>
          <ForemanModal
            isOpen={modalOpen}
            title="I'm a modal!"
            onClose={toggleModal}
            myProp="Hii"
          >
            <ForemanModal.Header />
            The inner {`<Modal>`} component will have any props you pass to
            {`<ForemanModal>`}. (Look in the React dev tools for
            &lsquo;myProp&rsquo;)
            <ForemanModal.Footer />
          </ForemanModal>
        </Story>
      );
    })
);
