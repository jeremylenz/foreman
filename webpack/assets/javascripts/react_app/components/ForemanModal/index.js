import React from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import { selectIsModalOpen } from './ForemanModalSelectors';
import { setModalClosed } from './ForemanModalActions';
import ForemanModal from './ForemanModal';
import ForemanModalHeader from './subcomponents/ForemanModalHeader';
import ForemanModalFooter from './subcomponents/ForemanModalFooter';
import reducer from './ForemanModalReducer';

export const reducers = { foremanModals: reducer };

const ConnectedForemanModal = props => {
  const { id, title } = props;
  const isOpen = useSelector(state => selectIsModalOpen(state, id));
  const dispatch = useDispatch();
  const onClose = () => dispatch(setModalClosed({ id }));
  return (
    <ForemanModal
      {...props}
      id={id}
      title={title}
      isOpen={isOpen}
      onClose={onClose}
    />
  );
};

ConnectedForemanModal.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string,
};

ConnectedForemanModal.defaultProps = {
  title: '',
};

// Header and Footer use the provided children, or default markup if none provided

ConnectedForemanModal.Header = ForemanModalHeader;
ConnectedForemanModal.Footer = ForemanModalFooter;

export default ConnectedForemanModal;
