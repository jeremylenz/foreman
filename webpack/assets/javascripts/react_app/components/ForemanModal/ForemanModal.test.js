import React from 'react';
import ForemanModal from '.';
import { testComponentSnapshotsWithFixtures } from '../../common/testHelpers';

const headerChild = (
  <ForemanModal.Header>this is the header</ForemanModal.Header>
);
const footerChild = (
  <ForemanModal.Footer>this is the footer</ForemanModal.Footer>
);
const modalBody = <div>This is the body</div>;

const fixtures = {
  'should render': {
    isOpen: true,
    title: 'Test modal',
    onClose: jest.fn(),
  },
  'should render closed': {
    isOpen: false,
    title: 'Test modal',
    onClose: jest.fn(),
  },
  'renders when header and footer are supplied': {
    isOpen: true,
    title: 'Test modal',
    onClose: jest.fn(),
    children: [headerChild, modalBody, footerChild],
  },
  'renders header and footer in correct order regardless of ordering of children': {
    isOpen: true,
    title: 'Test modal',
    onClose: jest.fn(),
    children: [modalBody, footerChild, headerChild],
  },
};

describe('ForemanModal', () => {
  describe('rendering', () => {
    testComponentSnapshotsWithFixtures(ForemanModal, fixtures);
  });
});
