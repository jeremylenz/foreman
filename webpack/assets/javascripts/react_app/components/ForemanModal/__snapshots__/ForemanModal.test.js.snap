// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForemanModal rendering renders header and footer in correct order regardless of ordering of children 1`] = `
<ContextProvider
  value={
    Object {
      "isOpen": true,
      "onClose": [MockFunction],
      "title": "Test modal",
    }
  }
>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    className="foreman-modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[MockFunction]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={true}
  >
    <ForemanModalHeader
      key=".2"
    >
      this is the header
    </ForemanModalHeader>
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    >
      <div
        key=".0"
      >
        This is the body
      </div>
    </ModalBody>
    <ForemanModalFooter
      key=".1"
    >
      this is the footer
    </ForemanModalFooter>
  </Modal>
</ContextProvider>
`;

exports[`ForemanModal rendering renders when header and footer are supplied 1`] = `
<ContextProvider
  value={
    Object {
      "isOpen": true,
      "onClose": [MockFunction],
      "title": "Test modal",
    }
  }
>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    className="foreman-modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[MockFunction]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={true}
  >
    <ForemanModalHeader
      key=".0"
    >
      this is the header
    </ForemanModalHeader>
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    >
      <div
        key=".1"
      >
        This is the body
      </div>
    </ModalBody>
    <ForemanModalFooter
      key=".2"
    >
      this is the footer
    </ForemanModalFooter>
  </Modal>
</ContextProvider>
`;

exports[`ForemanModal rendering should render 1`] = `
<ContextProvider
  value={
    Object {
      "isOpen": true,
      "onClose": [MockFunction],
      "title": "Test modal",
    }
  }
>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    className="foreman-modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[MockFunction]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={true}
  >
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    />
  </Modal>
</ContextProvider>
`;

exports[`ForemanModal rendering should render closed 1`] = `
<ContextProvider
  value={
    Object {
      "isOpen": false,
      "onClose": [MockFunction],
      "title": "Test modal",
    }
  }
>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    className="foreman-modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[MockFunction]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={false}
  >
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    />
  </Modal>
</ContextProvider>
`;
